File: netlify/functions/gerar-pdf.py
import json
import math
import base64
import tempfile
import os
from datetime import date as _date
from reportlab.lib.pagesizes import A4
from reportlab.lib import colors
from reportlab.lib.units import cm
from reportlab.pdfgen import canvas


def euros_to_kms(v, valor_km):
    return math.ceil(v / valor_km)


PORTO_DISTANCES = [
    ("Matosinhos", 17), ("Vila Nova de Gaia", 22), ("Espinho", 40),
    ("Maia", 25), ("Valongo", 30), ("Paredes", 60), ("Famalicão", 70),
    ("Penafiel", 70), ("Felgueiras", 80), ("Barcelos", 90),
    ("Guimarães", 100), ("Braga", 110), ("Amarante", 120),
    ("Viana do Castelo", 140), ("Vila Real", 160), ("Aveiro", 130),
    ("Coimbra", 240), ("Viseu", 260), ("Chaves", 320), ("Bragança", 450),
]


def km_to_city(kms):
    return min(PORTO_DISTANCES, key=lambda x: abs(x[1] - kms))[0]


def generate_pdf(payload):
    func     = payload["funcionaria"]
    empresa  = payload["empresa"]
    mes      = payload["mes"]
    ano      = payload["ano"]
    valor_km = payload.get("valor_km", 0.40)

    parques_como_km = []
    for p in payload.get("parques", []):
        kms_eq = euros_to_kms(p["valor"], valor_km)
        parques_como_km.append({
            "dia": p["dia"], "origem": "Porto",
            "destino": km_to_city(kms_eq),
            "just": p.get("evento", ""),
            "kms": kms_eq, "_parque": True
        })

    kms_raw = [dict(r) for r in payload.get("kms", [])]
    for r in kms_raw:
        if r.get("origem", "Porto").strip().lower() != "porto":
            r["destino"] = km_to_city(r["kms"])
        r["origem"] = "Porto"

    kms_por_dia = {}
    for i, r in enumerate(kms_raw):
        if r["dia"] not in kms_por_dia:
            kms_por_dia[r["dia"]] = i

    parques_standalone = []
    for p in parques_como_km:
        if p["dia"] in kms_por_dia:
            idx = kms_por_dia[p["dia"]]
            kms_raw[idx]["kms"] += p["kms"]
        else:
            parques_standalone.append(p)

    todas = kms_raw + parques_standalone
    todas.sort(key=lambda x: x["dia"])
    dias_usados = {r["dia"]: r for r in todas}
    total_kms   = sum(r["kms"] for r in todas)
    total_valor = total_kms * valor_km

    tmp = tempfile.NamedTemporaryFile(suffix=".pdf", delete=False)
    output_path = tmp.name
    tmp.close()

    W, H = A4
    ML = 1.5 * cm
    MT = 1.4 * cm

    DARK      = colors.HexColor("#1a1a1a")
    MID       = colors.HexColor("#666666")
    BORDER    = colors.HexColor("#cccccc")
    BORDER_LT = colors.HexColor("#e5e5e5")
    HEADER_BG = colors.HexColor("#2c2c2c")
    ROW_ALT   = colors.HexColor("#f7f7f7")
    PARQUE_BG = colors.HexColor("#eef4fb")
    BG1       = colors.HexColor("#f0f0f0")
    BG2       = colors.white

    c = canvas.Canvas(output_path, pagesize=A4)
    c.setTitle(f"Mapa de Deslocações - {func['nome']} - {mes} {ano}")

    y = H - MT

    def lbl(cx, cy, txt):
        c.setFont("Helvetica-Bold", 6.5)
        c.setFillColor(MID)
        c.drawString(cx, cy, txt)

    def val(cx, cy, txt, bold=False, size=8.5):
        c.setFont("Helvetica-Bold" if bold else "Helvetica", size)
        c.setFillColor(DARK)
        c.drawString(cx, cy, txt)

    def hline(yy, weight=0.4, clr=None):
        c.setStrokeColor(clr or BORDER)
        c.setLineWidth(weight)
        c.line(ML, yy, W - ML, yy)

    # TÍTULO
    c.setFont("Helvetica-Bold", 15)
    c.setFillColor(DARK)
    c.drawString(ML, y, "Mapa de Deslocações")
    y -= 0.55 * cm
    hline(y, weight=1.6, clr=DARK)
    y -= 0.40 * cm

    ROW_H = 0.48 * cm
    PAD_X = 0.25 * cm
    PAD_Y = 0.13 * cm
    TW  = W - 2*ML
    cL1 = ML
    cV1 = ML + 2.2*cm
    cL2 = ML + TW/2
    cV2 = ML + TW/2 + 2.2*cm

    def header_row(yy, lbl1, val1, lbl2, val2, bg=BG1):
        c.setFillColor(bg)
        c.rect(ML, yy - ROW_H + 0.10*cm, TW, ROW_H, fill=1, stroke=0)
        c.setStrokeColor(BORDER); c.setLineWidth(0.3)
        c.line(ML, yy - ROW_H + 0.10*cm, ML + TW, yy - ROW_H + 0.10*cm)
        ty = yy - ROW_H + PAD_Y + 0.10*cm
        c.setFont("Helvetica-Bold", 8); c.setFillColor(DARK)
        c.drawString(cL1 + PAD_X, ty, lbl1)
        if lbl2:
            c.drawString(cL2 + PAD_X, ty, lbl2)
        c.setFont("Helvetica", 8)
        c.drawString(cV1 + PAD_X, ty, val1)
        if val2:
            c.drawString(cV2 + PAD_X, ty, val2)
        return yy - ROW_H

    y = header_row(y, "Empresa", empresa["nome"], "Ano", ano, bg=BG1)
    y = header_row(y, "Morada",  empresa["morada"], "Mês", mes, bg=BG2)
    y = header_row(y, "NIF",     empresa["nif"], "", "", bg=BG1)
    y -= 0.28 * cm

    c.setFont("Helvetica", 8); c.setFillColor(DARK)
    c.drawString(ML, y, "Recebido por:")
    y -= 0.30 * cm

    y = header_row(y, "Nome",   func["nome"],   "Viatura", func["viatura"], bg=BG1)
    y = header_row(y, "Morada", func["morada"], "Data",    _date.today().strftime("%d/%m/%Y"), bg=BG2)

    c.setFillColor(BG1)
    c.rect(ML, y - ROW_H + 0.10*cm, TW, ROW_H, fill=1, stroke=0)
    c.setStrokeColor(BORDER); c.setLineWidth(0.3)
    c.line(ML, y - ROW_H + 0.10*cm, ML + TW, y - ROW_H + 0.10*cm)
    ty = y - ROW_H + PAD_Y + 0.10*cm
    c.setFont("Helvetica-Bold", 8); c.setFillColor(DARK)
    c.drawString(cL1 + PAD_X, ty, "NIF")
    c.drawString(cL2 + PAD_X, ty, "Assinatura")
    c.setFont("Helvetica", 8)
    c.drawString(cV1 + PAD_X, ty, func["nif"])
    c.setStrokeColor(BORDER); c.setLineWidth(0.4)
    c.line(cV2 + PAD_X, ty - 0.02*cm, ML + TW - PAD_X, ty - 0.02*cm)
    y -= ROW_H
    y -= 0.32 * cm

    hline(y)
    y -= 0.32 * cm

    tot_w = TW / 3
    tA = ML; tB = ML + tot_w; tC = ML + tot_w * 2

    c.setFillColor(BG2)
    c.rect(ML, y - ROW_H*2 + 0.10*cm, TW, ROW_H*2, fill=1, stroke=0)
    ty1 = y - ROW_H*0.45
    ty2 = y - ROW_H*1.4

    c.setFont("Helvetica", 7.5); c.setFillColor(MID)
    c.drawString(tA + PAD_X, ty1, "Valor por KM, conforme portaria")
    c.drawString(tB + PAD_X, ty1, "Total KMs")
    c.setFont("Helvetica-Bold", 8); c.setFillColor(DARK)
    c.drawString(tC + PAD_X, ty1, "Total Recebido")
    c.setFont("Helvetica", 9); c.setFillColor(DARK)
    c.drawString(tA + PAD_X, ty2, f"{valor_km:.2f}\u20ac")
    c.drawString(tB + PAD_X, ty2, f"{total_kms}")
    c.setFont("Helvetica-Bold", 9)
    c.drawString(tC + PAD_X, ty2, f"{total_valor:.2f}\u20ac")
    y -= ROW_H * 2
    y -= 0.28 * cm

    hline(y)
    y -= 0.28 * cm

    if parques_standalone:
        c.setFont("Helvetica-Oblique", 6.5); c.setFillColor(MID)
        c.drawString(ML, y, "* Estacionamentos convertidos em KMs equivalentes (Porto \u2192 cidade mais próxima)")
        y -= 0.28 * cm

    col_widths = [0.85*cm, 2.2*cm, 2.8*cm, 5.95*cm, 1.7*cm]
    total_w    = sum(col_widths)
    headers    = ["DIA", "ORIGEM", "DESTINO", "JUSTIFICAÇÃO", "Nº KMs"]

    FOOTER_H = 0.68 * cm
    HH       = 0.48 * cm
    available = y - ML - FOOTER_H - HH
    RH = available / 31

    c.setFillColor(HEADER_BG)
    c.rect(ML, y - HH, total_w, HH, fill=1, stroke=0)
    c.setFont("Helvetica-Bold", 7.5); c.setFillColor(colors.white)
    x = ML
    for h, w in zip(headers, col_widths):
        ty = y - HH * 0.5 - 0.10*cm
        if h == "Nº KMs":
            c.drawRightString(x + w - 0.18*cm, ty, h)
        else:
            c.drawString(x + 0.16*cm, ty, h)
        x += w
    y -= HH

    for dia in range(1, 32):
        row = dias_usados.get(dia)
        is_parque = row and row.get("_parque")

        if is_parque:
            c.setFillColor(PARQUE_BG)
            c.rect(ML, y - RH, total_w, RH, fill=1, stroke=0)
        elif dia % 2 == 0:
            c.setFillColor(ROW_ALT)
            c.rect(ML, y - RH, total_w, RH, fill=1, stroke=0)

        c.setStrokeColor(BORDER_LT); c.setLineWidth(0.25)
        c.line(ML, y - RH, ML + total_w, y - RH)

        ty = y - RH * 0.5 - 0.10*cm
        row_vals = [
            str(dia),
            row["origem"]   if row else "",
            row["destino"]  if row else "",
            row["just"]     if row else "",
            str(row["kms"]) if row else "",
        ]

        x = ML
        for i, (v, w) in enumerate(zip(row_vals, col_widths)):
            if i == 0:
                c.setFont("Helvetica-Bold", 7.5)
                c.setFillColor(DARK if row else BORDER)
                c.drawString(x + 0.16*cm, ty, v)
            elif i == 4:
                c.setFont("Helvetica-Bold" if row else "Helvetica", 7.5)
                c.setFillColor(DARK)
                c.drawRightString(x + w - 0.18*cm, ty, v)
            else:
                c.setFont("Helvetica-Oblique" if is_parque else "Helvetica", 7.5)
                c.setFillColor(DARK if row else BORDER)
                c.drawString(x + 0.16*cm, ty, v)
            x += w
        y -= RH

    hline(y)
    y -= 0.26 * cm
    c.setFont("Helvetica-Bold", 8); c.setFillColor(DARK)
    c.drawString(ML, y, f"Total KMs: {total_kms}")
    c.drawString(ML + 3.6*cm, y, f"Total a receber: {total_valor:.2f}\u20ac")
    c.drawRightString(W - ML, y, f"Neva Agency \u2014 {mes} {ano}")

    c.save()

    with open(output_path, "rb") as f:
        pdf_bytes = f.read()
    os.unlink(output_path)
    return pdf_bytes, func["nome"], mes, ano


def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return {
            "statusCode": 200,
            "headers": {
                "Access-Control-Allow-Origin": "*",
                "Access-Control-Allow-Headers": "Content-Type",
                "Access-Control-Allow-Methods": "POST, OPTIONS",
            },
            "body": ""
        }

    try:
        payload = json.loads(event.get("body", "{}"))
        pdf_bytes, nome, mes, ano = generate_pdf(payload)
        filename = f"Deslocacoes_{nome.split()[0]}_{mes}_{ano}.pdf"
        return {
            "statusCode": 200,
            "headers": {
                "Content-Type": "application/pdf",
                "Content-Disposition": f'attachment; filename="{filename}"',
                "Access-Control-Allow-Origin": "*",
            },
            "body": base64.b64encode(pdf_bytes).decode("utf-8"),
            "isBase64Encoded": True,
        }
    except Exception as e:
        return {
            "statusCode": 500,
            "headers": {"Access-Control-Allow-Origin": "*"},
            "body": json.dumps({"error": str(e)})
        }

File: netlify/functions/gerar-pdf.js
const { PDFDocument, rgb, StandardFonts } = require('pdf-lib');

const PORTO_DISTANCES = [
  ["Matosinhos", 17], ["Vila Nova de Gaia", 22], ["Espinho", 40],
  ["Maia", 25], ["Valongo", 30], ["Paredes", 60], ["Famalicão", 70],
  ["Penafiel", 70], ["Felgueiras", 80], ["Barcelos", 90],
  ["Guimarães", 100], ["Braga", 110], ["Amarante", 120],
  ["Viana do Castelo", 140], ["Vila Real", 160], ["Aveiro", 130],
  ["Coimbra", 240], ["Viseu", 260], ["Chaves", 320], ["Bragança", 450],
];

function kmToCity(kms) {
  return PORTO_DISTANCES.reduce((best, cur) =>
    Math.abs(cur[1] - kms) < Math.abs(best[1] - kms) ? cur : best
  )[0];
}

function eurosToKms(v, valorKm) {
  return Math.ceil(v / valorKm);
}

function today() {
  const d = new Date();
  return `${String(d.getDate()).padStart(2,'0')}/${String(d.getMonth()+1).padStart(2,'0')}/${d.getFullYear()}`;
}

async function generatePDF(payload) {
  const func    = payload.funcionaria;
  const empresa = payload.empresa;
  const mes     = payload.mes;
  const ano     = payload.ano;
  const valorKm = payload.valor_km || 0.40;

  // Build parking entries
  const parquesComoKm = (payload.parques || []).map(p => {
    const kmsEq = eurosToKms(p.valor, valorKm);
    return { dia: p.dia, origem: 'Porto', destino: kmToCity(kmsEq), just: p.evento || '', kms: kmsEq, _parque: true };
  });

  // Normalize KM entries: always Porto origin
  const kmsRaw = (payload.kms || []).map(r => {
    const entry = { ...r };
    if ((entry.origem || 'Porto').trim().toLowerCase() !== 'porto') {
      entry.destino = kmToCity(entry.kms);
    }
    entry.origem = 'Porto';
    return entry;
  });

  // Build day index (first entry per day)
  const kmsPorDia = {};
  kmsRaw.forEach((r, i) => { if (!(r.dia in kmsPorDia)) kmsPorDia[r.dia] = i; });

  // Merge parking into same-day trip or keep standalone
  const parquesStandalone = [];
  parquesComoKm.forEach(p => {
    if (p.dia in kmsPorDia) {
      kmsRaw[kmsPorDia[p.dia]].kms += p.kms;
    } else {
      parquesStandalone.push(p);
    }
  });

  const todas = [...kmsRaw, ...parquesStandalone].sort((a, b) => a.dia - b.dia);
  const diasUsados = {};
  todas.forEach(r => { diasUsados[r.dia] = r; });
  const totalKms   = todas.reduce((s, r) => s + r.kms, 0);
  const totalValor = totalKms * valorKm;

  // ── PDF setup ──────────────────────────────────────────────────────────────
  const pdfDoc = await PDFDocument.create();
  const page   = pdfDoc.addPage([595.28, 841.89]); // A4
  const { width, height } = page.getSize();

  const fontR = await pdfDoc.embedFont(StandardFonts.Helvetica);
  const fontB = await pdfDoc.embedFont(StandardFonts.HelveticaBold);
  const fontI = await pdfDoc.embedFont(StandardFonts.HelveticaOblique);

  const ML = 42; // margin left/right (pts)
  const MT = 40; // margin top
  const TW = width - ML * 2;

  const DARK    = rgb(0.10, 0.09, 0.09);
  const MID     = rgb(0.40, 0.40, 0.40);
  const LIGHT   = rgb(0.63, 0.63, 0.63);
  const BG1     = rgb(0.94, 0.94, 0.94);
  const BG2     = rgb(1, 1, 1);
  const HBG     = rgb(0.17, 0.17, 0.17);
  const ROW_ALT = rgb(0.97, 0.97, 0.97);
  const BORD    = rgb(0.80, 0.80, 0.80);
  const BORD_LT = rgb(0.90, 0.90, 0.90);
  const PARK_BG = rgb(0.93, 0.96, 0.98);

  let y = height - MT;

  function drawText(text, x, yy, { font = fontR, size = 8, color = DARK } = {}) {
    page.drawText(String(text), { x, y: yy, size, font, color });
  }

  function drawRect(x, yy, w, h, color) {
    page.drawRectangle({ x, y: yy, width: w, height: h, color, borderWidth: 0 });
  }

  function hline(yy, thickness = 0.4, color = BORD) {
    page.drawLine({ start: { x: ML, y: yy }, end: { x: ML + TW, y: yy }, thickness, color });
  }

  // ── TITLE ──────────────────────────────────────────────────────────────────
  drawText('Mapa de Deslocações', ML, y, { font: fontB, size: 15, color: DARK });
  y -= 16;
  hline(y, 1.6, DARK);
  y -= 11;

  // ── HEADER TABLE ───────────────────────────────────────────────────────────
  const ROW_H = 13.6;
  const PAD_X = 7;
  const cL1   = ML;
  const cV1   = ML + 62;
  const SPLIT = TW * 0.62;
  const cL2   = ML + SPLIT;
  const cV2   = ML + SPLIT + 62;

  function headerRow(yy, label1, value1, label2, value2, bg) {
    drawRect(ML, yy - ROW_H + 3, TW, ROW_H, bg);
    page.drawLine({ start: { x: ML, y: yy - ROW_H + 3 }, end: { x: ML + TW, y: yy - ROW_H + 3 }, thickness: 0.3, color: BORD });
    const ty = yy - ROW_H + 5.5;
    drawText(label1, cL1 + PAD_X, ty, { font: fontB, size: 8, color: DARK });
    drawText(value1, cV1 + PAD_X, ty, { font: fontR, size: 8, color: DARK });
    if (label2) drawText(label2, cL2 + PAD_X, ty, { font: fontB, size: 8, color: DARK });
    if (value2) drawText(value2, cV2 + PAD_X, ty, { font: fontR, size: 8, color: DARK });
    return yy - ROW_H;
  }

  y = headerRow(y, 'Empresa', empresa.nome, 'Ano', ano, BG1);
  y = headerRow(y, 'Morada',  empresa.morada, 'Mês', mes, BG2);
  y = headerRow(y, 'NIF',     empresa.nif, '', '', BG1);
  y -= 8;

  drawText('Recebido por:', ML, y, { font: fontR, size: 8, color: DARK });
  y -= 8;

  y = headerRow(y, 'Nome',   func.nome,   'Viatura', func.viatura, BG1);
  y = headerRow(y, 'Morada', func.morada, 'Data',    today(),      BG2);

  // NIF + assinatura row — taller if we have a signature image
  const SIG_H = func.assinaturaBase64 ? ROW_H * 3.5 : ROW_H;
  drawRect(ML, y - SIG_H + 3, TW, SIG_H, BG1);
  page.drawLine({ start: { x: ML, y: y - SIG_H + 3 }, end: { x: ML + TW, y: y - SIG_H + 3 }, thickness: 0.3, color: BORD });
  const ty3 = y - ROW_H + 5.5;
  drawText('NIF',        cL1 + PAD_X, ty3, { font: fontB, size: 8, color: DARK });
  drawText(func.nif,     cV1 + PAD_X, ty3, { font: fontR, size: 8, color: DARK });
  drawText('Assinatura', cL2 + PAD_X, ty3, { font: fontB, size: 8, color: DARK });
  if (func.assinaturaBase64) {
    try {
      const sigBytes = Buffer.from(func.assinaturaBase64, 'base64');
      const sigImg   = await pdfDoc.embedPng(sigBytes);
      const sigDims  = sigImg.scaleToFit(ML + TW - PAD_X - (cV2 + PAD_X), SIG_H - 6);
      page.drawImage(sigImg, {
        x: cV2 + PAD_X,
        y: y - SIG_H + 5,
        width:  sigDims.width,
        height: sigDims.height,
      });
    } catch(e) {
      // fallback: just draw the line if image fails
      page.drawLine({ start: { x: cV2 + PAD_X, y: ty3 - 1 }, end: { x: ML + TW - PAD_X, y: ty3 - 1 }, thickness: 0.4, color: BORD });
    }
  } else {
    page.drawLine({ start: { x: cV2 + PAD_X, y: ty3 - 1 }, end: { x: ML + TW - PAD_X, y: ty3 - 1 }, thickness: 0.4, color: BORD });
  }
  y -= SIG_H;
  y -= 9;

  hline(y);
  y -= 9;

  // ── TOTALS ─────────────────────────────────────────────────────────────────
  const tW3 = TW / 3;
  const tA  = ML, tB2 = ML + tW3, tC = ML + tW3 * 2;

  drawRect(ML, y - ROW_H * 2 + 3, TW, ROW_H * 2, BG2);
  const ty1 = y - ROW_H * 0.5;
  const ty2 = y - ROW_H * 1.5;

  drawText('Valor por KM, conforme portaria', tA + PAD_X, ty1, { font: fontR, size: 7.5, color: MID });
  drawText('Total KMs',                       tB2 + PAD_X, ty1, { font: fontR, size: 7.5, color: MID });
  drawText('Total Recebido',                  tC + PAD_X,  ty1, { font: fontB, size: 8,   color: DARK });

  drawText(`${valorKm.toFixed(2)}€`, tA + PAD_X,  ty2, { font: fontR, size: 9, color: DARK });
  drawText(`${totalKms}`,            tB2 + PAD_X, ty2, { font: fontR, size: 9, color: DARK });
  drawText(`${totalValor.toFixed(2)}€`, tC + PAD_X, ty2, { font: fontB, size: 9, color: DARK });

  y -= ROW_H * 2;
  y -= 8;
  hline(y);
  y -= 7;

  if (parquesStandalone.length) {
    drawText('* Estacionamentos convertidos em KMs equivalentes (Porto → cidade mais próxima)', ML, y, { font: fontI, size: 6.5, color: MID });
    y -= 8;
  }

  // ── TABLE ──────────────────────────────────────────────────────────────────
  const _fixed = 24 + 62 + 79 + 48;
  const colW  = [24, 62, 79, TW - _fixed, 48];
  const totW  = colW.reduce((s, w) => s + w, 0);
  const heads = ['DIA', 'ORIGEM', 'DESTINO', 'JUSTIFICAÇÃO', 'Nº KMs'];
  const FOOTER_H = 19;
  const HH       = 13.6;
  const available = y - ML - FOOTER_H - HH;
  const RH = available / 31;

  // Header bar
  drawRect(ML, y - HH, totW, HH, HBG);
  let x = ML;
  heads.forEach((h, i) => {
    const tx = i === 4 ? x + colW[i] - fontB.widthOfTextAtSize(h, 7.5) - 5 : x + 4.5;
    drawText(h, tx, y - HH + 4, { font: fontB, size: 7.5, color: rgb(1,1,1) });
    x += colW[i];
  });
  y -= HH;

  // 31 rows
  for (let dia = 1; dia <= 31; dia++) {
    const row      = diasUsados[dia];
    const isParque = row && row._parque;

    if (isParque) {
      drawRect(ML, y - RH, totW, RH, PARK_BG);
    } else if (dia % 2 === 0) {
      drawRect(ML, y - RH, totW, RH, ROW_ALT);
    }
    page.drawLine({ start: { x: ML, y: y - RH }, end: { x: ML + totW, y: y - RH }, thickness: 0.25, color: BORD_LT });

    const ty = y - RH / 2 - 3;
    const vals = [
      String(dia),
      row ? row.origem  : '',
      row ? row.destino : '',
      row ? row.just    : '',
      row ? String(row.kms) : '',
    ];

    x = ML;
    vals.forEach((v, i) => {
      const font  = (i === 0 || (i === 4 && row)) ? fontB : (isParque && i > 0 && i < 4 ? fontI : fontR);
      const color = (i === 0 && !row) ? LIGHT : (i > 0 && !row) ? LIGHT : DARK;
      if (i === 4 && v) {
        const tw = font.widthOfTextAtSize(v, 7.5);
        drawText(v, x + colW[i] - tw - 5, ty, { font, size: 7.5, color });
      } else {
        drawText(v, x + 4.5, ty, { font, size: 7.5, color });
      }
      x += colW[i];
    });
    y -= RH;
  }

  // ── FOOTER ─────────────────────────────────────────────────────────────────
  hline(y);
  y -= 7;
  drawText(`Total KMs: ${totalKms}`, ML, y, { font: fontB, size: 8, color: DARK });
  drawText(`Total a receber: ${totalValor.toFixed(2)}€`, ML + 102, y, { font: fontB, size: 8, color: DARK });
  const rightText = `Neva Agency — ${mes} ${ano}`;
  const rtw = fontB.widthOfTextAtSize(rightText, 8);
  drawText(rightText, ML + TW - rtw, y, { font: fontB, size: 8, color: DARK });

  const pdfBytes = await pdfDoc.save();
  return Buffer.from(pdfBytes);
}

exports.handler = async (event) => {
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
  };

  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 200, headers, body: '' };
  }

  try {
    const payload  = JSON.parse(event.body || '{}');
    const pdfBytes = await generatePDF(payload);
    const nome     = payload.funcionaria?.nome?.split(' ')[0] || 'Funcionaria';
    const filename = `Deslocacoes_${nome}_${payload.mes}_${payload.ano}.pdf`;
    return {
      statusCode: 200,
      headers: {
        ...headers,
        'Content-Type': 'application/pdf',
        'Content-Disposition': `attachment; filename="${filename}"`,
      },
      body: pdfBytes.toString('base64'),
      isBase64Encoded: true,
    };
  } catch (e) {
    console.error(e);
    return { statusCode: 500, headers, body: JSON.stringify({ error: e.message }) };
  }
};
